<launch>
    <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>

    <arg name="namespace"  default=""/>
    <arg name="x_pos" default="0.0"/>
    <arg name="y_pos" default="0.0"/>
    <arg name="z_pos" default="0.0"/>
    <arg name="yaw"   default="1.57"/>
    <arg name="move_forward_only" default="false"/>

    <group ns = "$(arg namespace)">
        <param name="robot_description" command="$(find xacro)/xacro --inorder $(find ros_sim)/description/urdf/turtlebot3_$(arg model).urdf.xacro" />
        <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg namespace) -x $(arg x_pos) -y $(arg y_pos) -z $(arg z_pos) -Y $(arg yaw) -param robot_description" />

        <!--robot_state_publisher, publishing state of robot to tf-->
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
        <param name="publish_frequency" type="double" value="50.0" />
        <param name="tf_prefix" value="$(arg namespace)" />
        </node>    

        <!-- robot pose publisher, publish the robot position on the map -->
        <node pkg="multi_robot_sim" type="robot_pose_publisher.py" name="robot_pose_publisher" args="$(arg namespace)" output="screen"/>

        <!-- status relay, publish robot status to a node -->
        <node pkg="multi_robot_sim" type="status_relay.py" name="status_relay" args="$(arg namespace)" output="screen"/>

        <!-- Waypoint navigation -->
        <node pkg="multi_robot_sim" type="waypoint_navigation.py" name="waypoint_nav" args="$(arg namespace)" output="screen"/>

        <!-- AMCL -->
        <include file="$(find multi_robot_sim)/launch/navigation/amcl_ns.launch">
        <arg name="robot_namespace" value="$(arg namespace)"/>
        <arg name="initial_pose_x" value="$(arg x_pos)"/>
        <arg name="initial_pose_y" value="$(arg y_pos)"/>
        <arg name="initial_pose_a" value="$(arg yaw)"/>

        </include>

        <!--move_base-->
        <include file="$(find multi_robot_sim)/launch/navigation/move_base_ns.launch">
            <arg name="model" value="$(arg model)" />
            <arg name="robot_namespace" value="$(arg namespace)"/>
            <arg name="move_forward_only" value="$(arg move_forward_only)"/>
        </include>  
    </group>
</launch>